src/Kestrel/KestrelConfigurationBuilder.cs

```diff
@@ -0,0 +1,306 @@
+        /// Specify callback to configure endpoint-specific settings.
+        /// </summary>
+        public KestrelConfigurationBuilder UnixSocketEndpoint(string socketPath, Action<ListenOptions> configure)
+        {
+            if (socketPath == null)
+            {
+                throw new ArgumentNullException(nameof(socketPath));
+            }
+            if (socketPath.Length == 0 || socketPath[0] != '/')
+            {
+                throw new ArgumentException(CoreStrings.UnixSocketPathMustBeAbsolute, nameof(socketPath));
+            }
+            if (configure == null)
+            {
+                throw new ArgumentNullException(nameof(configure));
+            }
+
+            EndpointsToAdd.Add(() =>
+            {
+                Options.ListenUnixSocket(socketPath, configure);
+            });
+
+            return this;
+        }
+
+        /// <summary>
+        /// Open a socket file descriptor.
+        /// </summary>
+        public KestrelConfigurationBuilder HandleEndpoint(ulong handle) => HandleEndpoint(handle, _ => { });
+
+        /// <summary>
+        /// Open a socket file descriptor.
+        /// The callback configures endpoint-specific settings.
+        /// </summary>
+        public KestrelConfigurationBuilder HandleEndpoint(ulong handle, Action<ListenOptions> configure)
+        {
+            if (configure == null)
+            {
+                throw new ArgumentNullException(nameof(configure));
+            }
+
+            EndpointsToAdd.Add(() =>
+            {
+                Options.ListenHandle(handle, configure);
+            });
+
+            return this;
+        }
+
+        public void Build()
+        {
+            if (Options.ConfigurationBuilder == null)
+            {
+                // The builder has already been built.
+                return;
+            }
+            Options.ConfigurationBuilder = null;
+
+            var configReader = new ConfigurationReader(Configuration);
+
+            LoadDefaultCert(configReader);
+            
+            foreach (var endpoint in configReader.Endpoints)
+            {
+                var listenOptions = AddressBinder.ParseAddress(endpoint.Url, out var https);
+                listenOptions.KestrelServerOptions = Options;
+                Options.EndpointDefaults(listenOptions);
+
+                var httpsOptions = new HttpsConnectionAdapterOptions();
+                if (https)
+                {
+                    // Defaults
+                    httpsOptions.ServerCertificate = listenOptions.KestrelServerOptions.GetOverriddenDefaultCertificate();
+                    Options.GetHttpsDefaults()(httpsOptions);
+                    if (httpsOptions.ServerCertificate == null)
+                    {
+                        var provider = Options.ApplicationServices.GetRequiredService<IDefaultHttpsProvider>();
+                        httpsOptions.ServerCertificate = provider.Certificate; // May be null.
```

Comment:
ConfigureHttps doesn't let you provide your own httpsOptions, or even a callback. The dev needs to be able to configure httpsOptions in configureEndpoint below before it's applied.

IDefaultHttpsProvider's signature is constrained by its use in Kestrel.Core, it can't expose any of the Https types directly.